### PR TITLE
Fix backend of XLA for Fused ConvBiasRelu

### PR DESCRIPTION
**Pull request by kaixih**
This PR is mainly to fix the issue of XLA sharing the same backend with native TF when calling fused ConvBiasRelu (both will call into `stream->FusedConvolveWithAlgorithm()`), which will cause segmentation fault problem when cudnn frontend APIs are used.

Therefore, this PR separate the `FusedConvolveWithAlgorithm` to (1) `FusedConvolveWithAlgorithm()` for legacy cudnn APIs and (2) `FusedConvolveWithExecutionPlan()` for cudnn frontend APIs. This behavior follows the `ConvolveWithAlgorithm()` and `ConvolveWithExecutionPlan()`.

Besides, the code has been greatly simplified by following the design of `ConvolveWithAlgorithm()` and removing many overloaded functions.

cc. @nluehr

### CODE REVIEW

**Comment by sanjoy** [comment]
@awpr is changing XLA to use the new cuDNN APIs as well.  Will this be PR be needed once that change lands?

**Comment by kaixih** [comment]
> @awpr is changing XLA to use the new cuDNN APIs as well. Will this be PR be needed once that change lands?

Curious how the XLA is going to support the new cuDNN's execution plan which is an opaque data structure at this moment. My understanding is that the execution plan (similar with algorithm index in the legacy cuDNN terms) needs to be serialized and our cuDNN+XLA team are working on that. @bas-aarts for vis.

Also, I think the PR is still needed since most of the PR is to simplify the code of FusedConvolveWithAlgorithm() to make it consistent with the design of ConvolveWithAlgorithm().

File: tensorflow/stream_executor/stream_executor_pimpl.h [hunk 1]
```diff
@@ -378,27 +378,15 @@ class StreamExecutor {
...
       const dnn::BatchDescriptor &input_descriptor,
       const dnn::FilterDescriptor &filter_descriptor,
       const dnn::BatchDescriptor &bias_descriptor,
       const dnn::BatchDescriptor &output_descriptor,
       const dnn::ConvolutionDescriptor &convolution_descriptor,
       std::vector<std::unique_ptr<dnn::ConvolveExecutionPlan>>
-          *out_exec_plans) {
-    dnn::DnnSupport *dnn_support = AsDnn();
-    if (dnn_support) {
-#if GOOGLE_CUDA
```
Comment (timshen91) [inline tensorflow/stream_executor/stream_executor_pimpl.h:391]: Please keep all GOOGLE_CUDA references in the headers, as GOOGLE_CUDA is often defined in the user land and propagated into the included headers (but not .cc files). Internally we rely on this kind of propagation.

**Review by timshen91 — APPROVED** [review]

**Comment by awpr** [comment]
My tentative plan was to make all the APIs internally look at the AlgorithmDesc to decide whether to use the Frontend API or the legacy entry point, and then switch over XLA by having it create an ExecutionPlan-based AlgorithmDesc instead of an enum-based one.  This goes in the opposite direction by having two separate entry points in the CudnnSupport for the two APIs.  I can probably work with either approach equally well.

As for deduplicating the DnnSupport methods, I was actually just looking into doing the same thing, so +1 to that.

Re: needing to serialize ExecutionPlans, I'm still looking into the details there, but my impression is that the backend API makes the entire underlying data available via `cudnnBackendGetAttribute`, so I should be able to painstakingly collect all the attributes (recursively for nested descriptor attributes), serialize that, and re-inflate the graph afterwards.  That won't work if the opaque structs behind the descriptor pointers contain data that's not visible via the attribute API, but I'm hopeful that's not the case, since the set of attributes looks at least plausibly sufficient to describe an underlying kernel launch.

File: tensorflow/stream_executor/cuda/cuda_dnn.cc [hunk 2]
```diff
@@ -4968,182 +4977,75 @@ port::Status CudnnSupport::DoBatchNormalizationBackwardImpl(
...
     return DoFusedConvolveImpl(
-        stream, conv_input_descriptor, conv_input_data, conv_input_scale,
+        stream, input_type, side_input_type, bias_type, output_type,
+        conv_input_descriptor, conv_input_data, conv_input_scale_float,
         filter_descriptor, filter_data, convolution_descriptor, side_input_data,
-        side_input_scale, bias_descriptor, biases, activation_mode,
+        side_input_scale_float, bias_descriptor, biases, activation_mode,
         output_descriptor, output_data,
-        GetConvAccumulatorType(dnn::DataType::kInt8), scratch_allocator,
+        GetConvAccumulatorType(dnn::DataType::kFloat), scratch_allocator,
```
[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/stream_executor/cuda/cuda_dnn.cc:5023]: That's a good catch. Fixing now.

File: tensorflow/stream_executor/rocm/rocm_dnn.h [hunk 3]
```diff
@@ -322,60 +322,18 @@ class MIOpenSupport : public dnn::DnnSupport {
       dnn::ProfileResult* output_profile_result) override;
 
   port::Status DoFusedConvolve(
```
[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/stream_executor/rocm/rocm_dnn.h:324]: I have no idea what's the plan of AMD. Should I delete them here or you guys will do that?

[the remark this answers is not in the document]
Comment (awpr) [reply tensorflow/stream_executor/rocm/rocm_dnn.h:324]: I see, this was added as part of the initial creation of the ROCm support; I was assuming ROCm support was older and it was added when adding support for fused convolutions.  If AMD devs added this override, it seems reasonable to leave it there.

**Comment by kaixih** [comment]
Just Note, I included two commits from https://github.com/tensorflow/tensorflow/pull/50216 to make sure the tests get passed. It resolved an issue related to the general use of frontend API. cc @timshen91

**Pull request merged**